core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,194 @@ angular.module('oppia').factory('ImprovementModalService', [
+              // a table in Multiple Incorrect Submissions issue.
+              let indicesToExpand = [];
+              let actionsToDisplay = [];
+
+              /**
+               * Identifies whether all learner actions have been displayed and
+               * no further expansion is possible.
+               * @returns {boolean}
+               */
+              $scope.canExpandActions = function() {
+                return currActionIndex !== -1;
+              }
+
+              /**
+               * Iterates through the learner actions and expands the set of
+               * actions to be displayed. If all learner actions have been
+               * handled, no more expansion is possible.
+               */
+              let expandActions = function() {
+                let i;
+                for (i = currActionIndex; i >= 0; i--) {
+                  let action = playthrough.actions[i];
+                  let currentStateName =
+                      action.actionCustomizationArgs.state_name.value;
+                  let latestStateName = actionsToDisplay.length === 0 ? (
+                      action.actionCustomizationArgs.state_name.value) : (
+                      actionsToDisplay[
+                        actionsToDisplay.length -
+                          1].actionCustomizationArgs.state_name.value);
+
+                  if (currentStateName !== latestStateName) {
+                    // When the latest state name and the state name of the
+                    // current learner action differ, they are part of
+                    // different contexts. So, we identify whether the size
+                    // of the block has crossed the threshold number of actions
+                    // per block. If it has, we don't add more learner actions
+                    // to the block.
+                    if (currActionIndex - i >= (
+                        UNRELATED_ACTIONS_PER_BLOCK_THRESHOLD)) {
+                      currActionIndex = i;
+                      break;
+                    }
+                    actionsToDisplay.push(action);
+                  } else {
+                    // When the latest state name and the state name of the
+                    // current learner action are the same, they remain in the
+                    // same context. So, we keep them in the same block to
+                    // display them together.
+                    actionsToDisplay.push(action);
+                  }
+                }
+                // This is the case when all learner actions have been iterated
+                // over, and no more expansion is possible.
+                if (i === -1) {
+                  currActionIndex = -1;
+                }
+                indicesToExpand.push(i + 1);
+              };
+
+              expandActions();
+
+              // Index to know where to start highlighting actions.
+              let finalBlockStartIndex = currActionIndex + 1;
```

Comment:
What does "final" mean in this context?

Reply:
So, this is sort of like something that happens on ngInit. After we call expandActions() the first time, we obtain the first block we display (which is technically the final block of the set of actions). So, on start, we also keep track of where the final block starts (in terms of learner action index), and highlight everything from that point onwards.

Reply:
Hm, ok, perhaps you could call it indexOfFirstBlockDisplayed instead.